python/graphit.py

```diff
@@ -0,0 +1,203 @@
+}
+
+class Graphit():
+
+    def __init__(self, username, password,
+                 client_id, client_secret,
+                 graph_endpoint, auth_endpoint,
+                 iam_endpoint=None):
+        self._headers = {'Content-type': 'application/json',
+                         'Accept': 'text/plain, application/json'
+                        }
+
+        self._username = username
+        self._password = password
+        self._client_id = client_id
+        self._client_secret = client_secret
+        self._graph_endpoint = graph_endpoint
+        self._auth_endpoint = auth_endpoint
+        self._iam_endpoint = iam_endpoint
+
+        self._token = None
+        self._token_expired = None
+
+    @backoff.on_exception(*BACKOFF_ARGS, **BACKOFF_KWARGS)
+    def get(self, url, token=None):
+        if token is None:
+            token, _token_expired = self.get_token()
+        headers = self._headers
+        if token is not None:
+            headers['Authorization'] = "Bearer " + token
+        res = requests.get(url, headers=headers, verify=False)
+        if res.status_code == 401:
+            self.refresh_token()
+            raise requests.exceptions.RequestException
+        try:
+            return json.loads(res.text)
+        except json.decoder.JSONDecodeError:
+            return res.text
+
+    @backoff.on_exception(*BACKOFF_ARGS, **BACKOFF_KWARGS)
+    def post(self, url, data, token=None):
+        if token is None:
+            token, _token_expired = self.get_token()
+        headers = self._headers
+        if token is not None:
+            headers['Authorization'] = "Bearer " + token
+        res = requests.post(url, data=json.dumps(data),
+                            headers=headers,
+                            verify=False)
+        if res.status_code == 401:
+            self.refresh_token()
+            raise requests.exceptions.RequestException
+        try:
+            return json.loads(res.text)
+        except json.decoder.JSONDecodeError:
+            return res.text
+
+    @backoff.on_exception(*BACKOFF_ARGS, **BACKOFF_KWARGS)
+    def delete(self, url, token=None):
+        if token is None:
+            token, _token_expired = self.get_token()
+        headers = self._headers
+        if token is not None:
+            headers['Authorization'] = "Bearer " + token
+        res = requests.delete(url, headers=headers, verify=False)
+        if res.status_code == 401:
+            self.refresh_token()
+            raise requests.exceptions.RequestException
+        try:
+            return json.loads(res.text)
+        except json.decoder.JSONDecodeError:
+            return res.text
+
+    def _timestamp(self):
+        return "[" + time.asctime(time.gmtime()) + " UTC]"
+
+    def _parse_response(self, res):
+        try:
+            body = json.loads(res.text)
+        except json.decoder.JSONDecodeError:
+            body = res.text
+        if 200 <= res.status_code <= 399:
+            return body
```

Comment:
this is inconsitsent error handling in most cases it would be "{'error':'text', 'code':123}" but in some cases it might be plain text? why?

Reply:
For this I'd like to refer to the discussion at mattermost corona-app on april 23 at 1PM. It was introduced as a fix to error handling, but there was never a concise specification how the library should react on the different cases. Especially the status code range 200 - 399 was taken from the java library.

In a nutshell: This point is definitely work in progress, needs discussion, and importantly a formal specification.